src/engine/SortedUnionImpl.h

```diff
@@ -0,0 +1,241 @@
+    localVocab_.mergeWith(std::span{&it->localVocab_, 1});
+    index = 0;
+    ++it;
+  }
+
+  // For the non-lazy case just append the remaining tables to the aggregated
+  // result table until the range is exhausted.
+  void appendRemaining(const std::vector<ColumnIndex>& permutation, auto& it,
+                       auto end, size_t& index) {
+    while (it != end) {
+      appendCurrent(permutation, it, index);
+    }
+  }
+
+  // Write a new row to the result table. The parameter `left` controls which
+  // permutation is used to write the row to the result table.
+  void pushRow(bool left, const auto& row) {
+    resultTable_.emplace_back();
+    for (size_t column = 0; column < resultTable_.numColumns(); column++) {
+      ColumnIndex origin = columnOrigins_.at(column).at(!left);
+      resultTable_.at(resultTable_.size() - 1, column) =
+          origin == Union::NO_COLUMN ? Id::makeUndefined() : row[origin];
+    }
+  }
+
+  // Increment the iterators if the current id table is fully processed and
+  // reset the index back to zero.
+  void advanceRangeIfConsumed() {
+    if (index1_ == it1_.value()->idTable_.size()) {
+      ++it1_.value();
+      index1_ = 0;
+    }
+    if (index2_ == it2_.value()->idTable_.size()) {
+      ++it2_.value();
+      index2_ = 0;
+    }
+  }
+
+  // Retrieve the current result from `resultTable_` and `localVocab_` and reset
+  // those members back to their initial value so the next operation can
+  // continue adding values.
+  Result::IdTableVocabPair popResult() {
+    auto result = Result::IdTableVocabPair{std::move(resultTable_),
+                                           std::move(localVocab_)};
+    resultTable_ = IdTable{resultTable_.numColumns(), allocator_};
+    resultTable_.reserve(Union::chunkSize);
+    localVocab_ = LocalVocab{};
+    return result;
+  }
+
+  // ___________________________________________________________________________
+  std::optional<Result::IdTableVocabPair> get() override {
+    if (aggregatedTableReturned_) {
+      return std::nullopt;
+    }
+    if (!it1_.has_value()) {
+      it1_ = range1_.begin();
+    }
+    if (!it2_.has_value()) {
+      it2_ = range2_.begin();
+    }
+    while (it1_ != range1_.end() && it2_ != range2_.end()) {
+      auto& idTable1 = it1_.value()->idTable_;
+      auto& idTable2 = it2_.value()->idTable_;
+      localVocab_.mergeWith(std::span{&it1_.value()->localVocab_, 1});
+      localVocab_.mergeWith(std::span{&it2_.value()->localVocab_, 1});
+      while (index1_ < idTable1.size() && index2_ < idTable2.size()) {
+        if (isSmaller(idTable1.at(index1_), idTable2.at(index2_))) {
+          pushRow(true, idTable1.at(index1_));
+          index1_++;
+        } else {
+          pushRow(false, idTable2.at(index2_));
+          index2_++;
+        }
+        if (requestLaziness_ && resultTable_.size() >= Union::chunkSize) {
+          auto result = popResult();
+          advanceRangeIfConsumed();
```

Comment:
I think this line can be simply dropped, as the next call to `get` will still handle this corner case correctly, saving us two lines. (simply  `return popResult()`).

Reply:
It would work correctly, but it would cause the local vocab of the previous table to get merged into the new local vocab unnecessarily